Model import

Enteli opened the pull request:
- [x] Read the README in ./Machine-Teaching-Examples/model_import
- [x] Reproduce each step in this sample
- [x] test importing from onnx format
- [x] test importing from tensorflow SavedModel format
- [x] Train Brain and verify that Moab can train to 100% success

PR song: https://music.youtube.com/watch?v=7grujCJntis&feature=share

Commit:
first

Commit:
first

journeyman-msft commented:
@Enteli Thank you for your service! I will avoid making notes on comments and stuff and stick to the functionality of what you provided. I can create a branch of write comments removing tutorial 1 for example.

journeyman-msft commented on Machine-Teaching-Examples/moab_control_type1a.ink:
```diff
@@ -0,0 +1,133 @@
+# To understand this Inkling better, please follow our tutorial walkthrough: 
+# https://aka.ms/moab/tutorial1
+
+#Control type 1a definition: immediate reward and single state dependent constraints: in this problem |x| < Squareside & |y| < Squareside
```
Was this Inkling snuck in?

Enteli replied:
Ah! my mistake, I had previously worked on a control_type story and thi wa sthe inkling. Was not meant to be part of this branch. removed

journeyman-msft commented on Machine-Teaching-Examples/model_import/environment.yml:
```diff
@@ -0,0 +1,12 @@
+  - scipy
+  - numpy
+  - tensorflow==1.15.2
+  - keras2onnx==1.7.0
```
```
- cryptography==2.9.2
- keras==2.4.2
- jupyter==1.0.0

Enteli replied:
ok

Enteli replied:
Added to yml

journeyman-msft reviewed (commented):
I was able to run through the onnx path using the jupyter notebook, import, and training. I was NOT to use the savedmodel builder in the jupyter notebook due to the following error:

```
FailedPreconditionError: Error while reading resource variable dense_1/bias from Container: localhost. This could mean that the variable was uninitialized. Not found: Container localhost does not exist. (Could not find resource: localhost/dense_1/bias)
```

Any idea why? I also had to add a few things to the environment.yml to get things to run, maybe Keras==2.4.2 is different than you are using?

onnx training is not 100%. Is that what you'd suspect since the state is 2x?
![image](https://user-images.githubusercontent.com/52263960/118690056-85b68c80-b7bc-11eb-9d00-10bba8cb176c.png)

Using the existing .zip
![image](https://user-images.githubusercontent.com/52263960/118690142-9bc44d00-b7bc-11eb-9f2e-741b8139361b.png)

Commit:
journey_resolved_comments1

Commit:
journey_comment2

Enteli commented:
> I was able to run through the onnx path using the jupyter notebook, import, and training. I was NOT to use the savedmodel builder in the jupyter notebook due to the following error:

I was able to identify the cause of the issue. The code needs  - keras==2.3.1. keras==2.4.2 is incompatible. fixed in environment.yml

Commit:
journey_comments3

Enteli commented:
>onnx training is not 100%. Is that what you'd suspect since the state is 2x?

- The model in the notebook is not the state*2x, it is a deep transform. 
- I reverted back to doing a state transform on state of 6 numbers (instead of 4). I verified that the brain trains and added a sample training graph in the README
- I think the deep transform on 4 number may not converge because the transform may squash the normalization (not sure why but not relevant for the purpose of demonstraing the functionality of model import)

journeyman-msft committed:
Added comments to Inkling, updated README with img

journeyman-msft committed:
Merge branch 'main' into model_import

journeyman-msft committed:
Upating README with training graph and bugfixes

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+# How To: build and import a tensorflow-keras model in bonsai
+
+This sample illustrates how to import a Machine Learning [ML] model in bonsai:
```
```suggestion
This sample illustrates how to import a Machine Learning [ML] model in Bonsai:
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+
+This sample illustrates how to import a Machine Learning [ML] model in bonsai:
+1. We build a ML model using tensorflow-keras
+   1. note: other frameworks such as pure tensorflow are supported, in this sample we focus on keras
```
```suggestion
   1. note: other frameworks such as pure Tensorflow are supported, in this sample we focus on Keras
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+1. We build a ML model using tensorflow-keras
+   1. note: other frameworks such as pure tensorflow are supported, in this sample we focus on keras
+   2. We illustrate by building a dummy deep transform:
+      1. A keras model (deep neural net) that transforms the input by applying a non linear transformation
```
```suggestion
      1. A Keras model (deep neural net) that transforms the input by applying a non linear transformation
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+2. Save the ML model as either
+   1. `tensorflow SavedModel` format OR
+   2. `onnx` format
+3. Import the saved model to the bonsai platform using bonsai-cli 
```
```suggestion
3. Import the saved model to the Bonsai platform using `bonsai-cli`
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+1. Create a conda virtual environment using the `environment.yml`.
+    `conda env create -f environment.yml`
+2. Currently only ML models compatible with tensorflow v1.15.2 are supported [(Link to docs)](https://docs.microsoft.com/en-us/bonsai/guides/import-ml-models?tabs=onnx%2Cnested1).
+3. When saving a keras model as a tensorflow SavedModel, one needs to save it using `tf.saved_model.builder.SavedModelBuilder()`. Saving a keras model using the direct `model.save()` is currently incompatible with bonsai model import. (see *state_transform_keras_model.ipynb* Jupyter notebook)
```
```suggestion
3. When saving a Keras model as a Tensorflow SavedModel, one needs to save it using `tf.saved_model.builder.SavedModelBuilder()`. Saving a Keras model using the direct `model.save()` is currently incompatible with Bonsai model import. (see *state_transform_keras_model.ipynb* Jupyter notebook)
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+1. Using the *state_transform_keras_model.ipynb* notebook
+   1. Build a tensorflow-keras model
+   2. Save the ML model as `tensorflow SavedModel` format OR `onnx` format
+2. Compressing the model as a `.zip`
```
```suggestion
2. Compress the model as a `.zip` (SavedModel only)
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+2. Compressing the model as a `.zip`
+   1. If the ML model is saved as a `tensorflow SavedModel`, compress the saved_model folder as a `.zip`
+   2. If the ML model is saved as `.onnx`, there is no need to compress as `.zip`
+3. Import the ML model to bonsai
```
```suggestion
3. Import the ML model to Bonsai
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+   1. If the ML model is saved as a `tensorflow SavedModel`, compress the saved_model folder as a `.zip`
+   2. If the ML model is saved as `.onnx`, there is no need to compress as `.zip`
+3. Import the ML model to bonsai
+   1. `tensorflow SavedModel` format OR
```
OR?

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
````diff
@@ -0,0 +1,65 @@
+            --modelfilepath ./state_transform_deep.zip 
+            --description "state transform tf SavedModel" 
+            --display-name <ml-model-name>
+        ```
````
Since we're giving a specific example, why not fill in the name and display name?

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
````diff
@@ -0,0 +1,65 @@
+            --modelfilepath ./state_transform_deep.onnx
+            --description "state transform ONNX" 
+            --display-name <ml-model-name>
+        ```
````
same -- fill in name, display name?

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
```diff
@@ -0,0 +1,65 @@
+ 1. Use the imported model to train your brain using a Moab sample
+    1. On preview.bons.ai
+       1. Create a new brain from a "Moab sample"
+       2. Copy paste the inkling from *moab-imported-concept.ink* in the *Teach* panel UI. In this sample we use the imported ML model to transform the state before passing it as a brain input (see inkling excerpt below).
```
```suggestion
       2. Copy paste the inkling from *moab-imported-concept.ink* in the *Teach* panel UI. In this sample we use the imported ML model to transform the state before passing it as a brain input, as follows:
```

shnayder commented on Machine-Teaching-Examples/model_import/README.md:
````diff
@@ -0,0 +1,65 @@
+       }
+       ```
+    2. Build the ML model as an imported concept named ImportedConcept: click `Build ImportedConcept` 
+    3. Trained the brain: click `Train MoveToCenter` (see below for an expected brain training graph) 
````
```suggestion
    3. Train the brain: click `Train MoveToCenter` (see below for an expected brain training graph) 
```

shnayder commented on Machine-Teaching-Examples/model_import/moab-imported-concept.ink:
```diff
@@ -0,0 +1,151 @@
+# Copyright 2021 Microsoft
+# This code is licensed under MIT license (see LICENSE for details)
+
+# Moab sample illustrating how to import a pre-tained neural network
```
```suggestion
# Moab sample illustrating how to import a pre-trained neural network
```

shnayder commented on Machine-Teaching-Examples/model_import/moab-imported-concept.ink:
```diff
@@ -0,0 +1,151 @@
+# This code is licensed under MIT license (see LICENSE for details)
+
+# Moab sample illustrating how to import a pre-tained neural network
+# model in to a bonsai brain solution. Use the jupyter notebook to
```
```suggestion
# model in a Bonsai brain. Use the Jupyter notebook to
```

shnayder commented on Machine-Teaching-Examples/model_import/moab-imported-concept.ink:
```diff
@@ -0,0 +1,151 @@
+# create a dummy deep state transform model in either Tensorflow v1.15.2 
+# SavedModelBuilder or ONNX format.
+
+# Use the bonsai-cli to import the model to the bonsai service:
```
```suggestion
# Use the bonsai-cli to import the model to Bonsai:
```

shnayder commented on Machine-Teaching-Examples/model_import/moab-imported-concept.ink:
```diff
@@ -0,0 +1,151 @@
+    # - must have an input state with the same dimesnions as the Inkling object
+    #   it maps to
+    concept ImportedConcept(input): ObservableState {
+        import {Model: "<ml-model-name>"} 
```
Ahh. Looks like you're actually using `"<ml-model-name>"` as the name. That's confusing. Perhaps "My ML model" or "State transform model" or something?

journeyman-msft replied:
Using "My ML Model" and updated top level comment to say `<ml-model-name>` might be "My ML Model"

shnayder commented on Machine-Teaching-Examples/model_import/moab-imported-concept.ink:
```diff
@@ -0,0 +1,151 @@
+    # Add the imported concept by name with the correct type definitions
+    # - can only have one input
+    # - cannot use image inputs
+    # - must have an input state with the same dimesnions as the Inkling object
```
```suggestion
    # - must have an input with the same dimensions as the Inkling concept
```

shnayder reviewed (approved):
Looks good -- suggested some spelling and capitalization and wording fixes.

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/README.md

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/README.md

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/README.md

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/README.md

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/README.md

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/moab-imported-concept.ink

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/moab-imported-concept.ink

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/moab-imported-concept.ink

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>

journeyman-msft committed:
Updating <ml-model-name> usage in Inkling

journeyman-msft committed:
Update Machine-Teaching-Examples/model_import/moab-imported-concept.ink

Co-authored-by: Victor Shnayder <victor.shnayder@microsoft.com>
